crates/sc-subspace-block-relay/src/consensus.rs

```diff
@@ -377,6 +441,56 @@ where
         Ok(response.encode())
     }
 
+    /// Handles the full download request from the client
+    fn on_full_download_request(
+        &mut self,
+        full_download_request: FullDownloadRequest<Block>,
+    ) -> Result<Vec<u8>, RelayError> {
+        let block_request = full_download_request.0;
+        let mut blocks = Vec::new();
+        let mut block_id = match block_request.from {
+            FromBlock::Hash(h) => BlockId::<Block>::Hash(h),
+            FromBlock::Number(n) => BlockId::<Block>::Number(n),
+        };
+
+        let mut total_size: usize = 0;
+        let max_blocks = block_request.max.map_or(MAX_RESPONSE_BLOCKS.into(), |val| {
+            std::cmp::min(val, MAX_RESPONSE_BLOCKS.into())
+        });
+        while blocks.len() < max_blocks as usize {
+            let block_hash = self.block_hash(&block_id)?;
+            let partial_block = self.get_partial_block(&block_hash, block_request.fields)?;
+            let body = if block_request.fields.contains(BlockAttributes::BODY) {
+                Some(block_transactions(&block_hash, self.client.as_ref())?)
+            } else {
+                None
+            };
+            let block_number = partial_block.block_number;
+            let parent_hash = partial_block.parent_hash;
+            let block_data = partial_block.block_data(body);
+
+            // Enforce the max limit on response size.
+            let bytes = block_data.encoded_size();
+            if !blocks.is_empty() && (total_size + bytes) > MAX_RESPONSE_SIZE.into() {
```

Comment:
This isn't 100% correct. Encoding depends on number of elements, `total_size + bytes` will not account for that. This is why I was recommending to push a new element, use `blocks.encoded_size()` from the beginning to check the size and popping last element if after insertion the limit was exceeded.

See https://docs.substrate.io/reference/scale-codec/#fn-1 for details of how compact numbers are encoded in SCALE codec.

Reply:
> But those few bytes might be enough for node to get stuck and being unable to continue sync because it is one or two bytes above the limit.

That is not how it works (e.g) say the block has one extrinsic which is > the limit, it will get sent (see similar check in upstream). The right solution for this would be chunking at the source. 

I don't mind making the change, but if there is a strong reason.

Reply:
> That is not how it works (e.g) say the block has one extrinsic which is > the limit, it will get sent (see similar check in upstream). The right solution for this would be chunking at the source.

I do not understand what you mean here. I think we have already established that upstream is buggy and shouldn't be followed. We also know that at least a single block will fit into the response, so we don't need to worry about that. Thus we don't need any chunking, we just send as many block as fit into the limit and client will reach out again if they need more blocks later.

Reply:
yeah I was concerned about the repeated `encoded_size()` calls(+ duplicate `encode()` in the end). Other option would be to keep the encoded `Vec<>` as we add entries, and use something like `EncodeAppend` to append. But the rollback of the last entry becomes expensive/messy.

I could not find anything satisfactory, so ended up implementing a simpler scheme. This allows appending encoded blocks, while able to check for overflow before append (no rollback needed)